[PyTorch] Add unit test for c10::Synchronized<T>

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74110
* #74109
* #74108
* #74107
* #74106
* #74105

Since this class was added, it's missing a unit test. @dbort first noticed it, so adding one in this commit.

Differential Revision: [D34822911](https://our.internmc.facebook.com/intern/diff/D34822911/)

dhruvbird committed:
[PyTorch] Add unit test for c10::Synchronized<T>

Since this class was added, it's missing a unit test. @dbort first noticed it, so adding one in this commit.

Differential Revision: [D34822911](https://our.internmc.facebook.com/intern/diff/D34822911/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Add unit test for c10::Synchronized<T>"


Since this class was added, it's missing a unit test. dbort first noticed it, so adding one in this commit.

Differential Revision: [D34822911](https://our.internmc.facebook.com/intern/diff/D34822911/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Add unit test for c10::Synchronized<T>"


Since this class was added, it's missing a unit test. dbort first noticed it, so adding one in this commit.

Differential Revision: [D34822911](https://our.internmc.facebook.com/intern/diff/D34822911/)

[ghstack-poisoned]

dbort commented on c10/test/util/Synchronized_test.cpp:
```diff
@@ -0,0 +1,43 @@
+    t.join();
+  }
+
+  iv.withLock([](int& iv) { EXPECT_EQ(iv, NUM_LOOP_INCREMENTS * 10); });
```
Consider using `threads.size()` instead of `10`

dbort reviewed (approved):
For posterity, looks like this is a second take of reverted PR https://github.com/pytorch/pytorch/pull/74062

dhruvbird committed:
Update on "[PyTorch] Add unit test for c10::Synchronized<T>"


Since this class was added, it's missing a unit test. dbort first noticed it, so adding one in this commit.

Differential Revision: [D34822911](https://our.internmc.facebook.com/intern/diff/D34822911/)

[ghstack-poisoned]
